### ISSUE TITLE
ips calculation impacted by model loading time

### ISSUE THREAD

**Issue by unphased** [papuSpartan/stable-diffusion-webui-distributed#11]
I have two 3090s, the slightly faster one (420w power limit) is on the slave instance and the master instance has the slower one (currently 300w power limit). 

Well, first, I encountered a repeated error 

```
webui-docker-auto-1  | Traceback (most recent call last):
webui-docker-auto-1  |   File "/stable-diffusion-webui/modules/call_queue.py", line 57, in f
webui-docker-auto-1  |     res = list(func(*args, **kwargs))
webui-docker-auto-1  |   File "/stable-diffusion-webui/modules/call_queue.py", line 37, in f
webui-docker-auto-1  |     res = func(*args, **kwargs)
webui-docker-auto-1  |   File "/stable-diffusion-webui/modules/txt2img.py", line 54, in txt2img
webui-docker-auto-1  |     processed = modules.scripts.scripts_txt2img.run(p, *args)
webui-docker-auto-1  |   File "/stable-diffusion-webui/modules/scripts.py", line 441, in run
webui-docker-auto-1  |     processed = script.run(p, *script_args)
webui-docker-auto-1  |   File "/stable-diffusion-webui/extensions/stable-diffusion-webui-distributed/scripts/extension.py", line 291, in run
webui-docker-auto-1  |     Script.world.optimize_jobs(payload)  # optimize work assignment before dispatching
webui-docker-auto-1  |   File "/stable-diffusion-webui/extensions/stable-diffusion-webui-distributed/scripts/spartan/World.py", line 402, in optimize_jobs
webui-docker-auto-1  |     lag = self.job_stall(job.worker, payload=payload)
webui-docker-auto-1  |   File "/stable-diffusion-webui/extensions/stable-diffusion-webui-distributed/scripts/spartan/World.py", line 330, in job_stall
webui-docker-auto-1  |     lag = worker.batch_eta(payload=payload, quiet=True) - fastest_worker.batch_eta(payload=payload, quiet=True)
webui-docker-auto-1  |   File "/stable-diffusion-webui/extensions/stable-diffusion-webui-distributed/scripts/spartan/Worker.py", line 210, in batch_eta
webui-docker-auto-1  |     eta = (num_images / self.avg_ipm) * 60
webui-docker-auto-1  | TypeError: unsupported operand type(s) for /: 'int' and 'NoneType'
```

I read this error message and got clued in that avg_ipm was None, so I manually ran the benchmark, and it resolved the issue.

However I noticed that with the benchmark my slave was getting a worse ipm rating: 

webui-docker-auto-1  |          1. 'master'(0.0.0.0:7860) - 51.36 ipm
webui-docker-auto-1  |          2. 'testrig_ftw3'(192.168.1.41:7860) - 39.91 ipm

This is surely because the slave instance is on a much inferior NVMe disk and i could watch as it was loading the model that it was way slower. 

Then in the course of actual generation of images and using the same model, no model loading wait is incurred, and both rigs crank out images at the same rate. 

I just wonder if we can address this somehow. Or maybe it's a non-issue... Regardless, even though they have quite different ipm ratings the image allocation isn't really being impacted (I tested with 32 images via 4 batches of 8 and the speed is awesome, 24 seconds).

**Comment by papuSpartan** [comment]
Glad you could get it working by forcing it to re-bench. Not exactly sure why the bench result is so far off for your other one. You might try editing to be a higher value which might improve the accuracy of the benchmark (though in doing this it will ofc take longer). https://github.com/papuSpartan/stable-diffusion-webui-distributed/blob/da2e32242eda19a15c787261a7c3a4dcebc631ed/scripts/spartan/shared.py#L10

Another thing you can do is run with `--distributed-debug` and you should be able to see each individual benchmark sample result. If you notice that they are vastly different than each other then you probably do just need to increase the warmup sample amount.

**Comment by unphased** [comment]
Thanks. Even on a 32 image generation it still did allocate 16 to each machine. Given those ipm numbers is this expected? Since it's actually allocating even amounts for some reason I have no reason to investigate further, but your notes will be very helpful for that. 

Working on bringing up my third computer for now. Later on once I get a bigger PSU i'll move the new 3090 into the workstation and go back to 2 machines.

**Comment by papuSpartan** [comment]
As you said, your generation time was ~24 seconds, so it does make sense that things were still distributed evenly and I'll explain why. At 51 ipm, 16 images should take about 18s, at 39.91 ipm, about 24s. The way distributed works is that it has to choose a relative "realtime", and I have set it to do so by picking the fastest worker. In this case, that would be master with 51 ipm. So seeing this, one's initial reaction might be "why isn't the slave generating slightly less?", that's because I have hardcoded an additional allowance of 6 seconds to account for network latency and other unknowns. This means that for the slave to no longer be considered realtime it would need to have an ETA of ~25s since 18.82 + 6 = 24.82.

Basically, I forgot to change job_timeout back to a more sensible value like 3 seconds in an earlier patch and so you're in a situation now where things will work just fine since ipm is utilized more as a rough ratio than an exact measure. Even in situations where the ETA is a bit longer it should swap over to complementary worker logic and assign almost the same amount to the slave since both of their ETA's are still very close. If you would like to prevent that from happening you could set job_timeout to a very high value in [the source](https://github.com/papuSpartan/stable-diffusion-webui-distributed/blob/f962b8dece3168e6ad95869e82a0c804a13016ce/scripts/spartan/World.py#L79C20-L79C20).

**Comment by papuSpartan** [comment]
I can fix this issue at some point though. I'll add it to the list.

**Comment by papuSpartan** [comment]
The model loading issue with benchmarking should be fixed by https://github.com/papuSpartan/stable-diffusion-webui-distributed/commit/a047ab832d24a73044237a0a156033626240906a. Reopen if needed
